pkg/dyninst/decode/decoder.go

```diff
@@ -0,0 +1,332 @@
+		out, err := d.parseDataToGoValue(iterator.addressPassMap, parameterType, parameterData)
+		if err != nil {
+			return fmt.Errorf("error parsing data for field %s: %w", eventRoot.Name, err)
+		}
+		err = d.uploader.WriteToken(jsontext.String(expr.Name))
+		if err != nil {
+			return err
+		}
+		err = json.MarshalEncode(d.uploader, out)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (d *Decoder) parseDataToGoValue(addressPassMap map[uint64]*dataItem, irType ir.Type, data []byte) (any, error) {
+	switch v := irType.(type) {
+	case *ir.BaseType:
+		return d.parseDataToGoBaseTypeValue(v, data)
+	case *ir.PointerType:
+		if len(data) < int(v.GetByteSize()) {
+			return nil, errors.New("passed data not long enough for pointer")
+		}
+		addr := binary.NativeEndian.Uint64(data)
+		pointedValue, ok := addressPassMap[addr]
+		if !ok {
+			return nil, errors.New("pointer not found in address pass map")
+		}
+		pointeeValue, err := d.parseDataToGoValue(addressPassMap, d.program.Types[ir.TypeID(pointedValue.header.Type)], pointedValue.data)
+		if err != nil {
+			return nil, fmt.Errorf("could not get pointed-at value: %s", err)
+		}
+		m := make(map[string]any, 1)
+		m[strconv.FormatUint(addr, 10)] = pointeeValue
+		return m, nil
+	case *ir.StructureType:
+		structFields := map[string]any{}
+		structure := irType.(*ir.StructureType)
+		for _, field := range structure.Fields {
+			v, err := d.parseDataToGoValue(addressPassMap, field.Type, data[field.Offset:field.Offset+uint32(field.Type.GetByteSize())])
+			if err != nil {
+				return nil, err
+			}
+			structFields[field.Name] = v
+		}
+		return structFields, nil
+	case *ir.ArrayType:
+		//TODO
+	case *ir.GoEmptyInterfaceType:
+		//TODO
+	case *ir.GoInterfaceType:
+		//TODO
+	case *ir.GoSliceHeaderType:
+		sliceHeader := irType.(*ir.GoSliceHeaderType)
+		var array, length, cap uint64
+		for _, field := range sliceHeader.Fields {
+			if field.Name == "array" {
+				if uint32(field.Type.GetByteSize()) != 8 || len(data) < int(field.Offset+field.Type.GetByteSize()) {
+					return nil, errors.New("malformed string field 'str'")
+				}
+				array = binary.NativeEndian.Uint64(data[field.Offset : field.Offset+uint32(field.Type.GetByteSize())])
+			} else if field.Name == "len" {
+				if field.Type.GetByteSize() != 8 {
+					return nil, errors.New("malformed string field 'len'")
+				}
+				length = binary.NativeEndian.Uint64(data[field.Offset : field.Offset+uint32(field.Type.GetByteSize())])
+			} else if field.Name == "cap" {
+				if field.Type.GetByteSize() != 8 {
+					return nil, errors.New("malformed string field 'len'")
+				}
+				cap = binary.NativeEndian.Uint64(data[field.Offset : field.Offset+uint32(field.Type.GetByteSize())])
+			} else {
+				fmt.Println(field.Name)
+			}
+		}
+
+		//TODO
+		fmt.Println(array, length, cap)
+
+	case *ir.GoSliceDataType:
+		//TODO
+	case *ir.GoChannelType:
+		//TODO
```

Comment:
let's return unimplemented errors and then test them explicitly -- or do the implementation